# Course project


**Основное**
- Дедлайн - 27 декабря 23:59
- Целевая метрика precision@5
- Бейзлайн решения - [MainRecommender](https://github.com/geangohn/recsys-tutorial/blob/master/src/recommenders.py)
- Сдаем ссылку на github с решением. В решении должны быть отчетливо видна метрика на новом тестовом сете из файла retail_test1.csv, то есть вам нужно для всех юзеров из этого файла выдать выши рекомендации, и посчитать на actual покупках precision@5. 

**!! Мы не рассматриваем холодный старт для пользователя, все наши пользователя одинаковы во всех сетах, поэтому нужно позаботиться об их исключении из теста.**


**Hints:** 

Сначала просто попробуйте разные параметры MainRecommender:  
- N в топ-N товарах при формировании user-item матирцы (сейчас топ-5000)  
- Различные веса в user-item матрице (0/1, кол-во покупок, log(кол-во покупок + 1), сумма покупки, ...)  
- Разные взвешивания матрицы (TF-IDF, BM25 - у него есть параметры)  
- Разные смешивания рекомендаций (обратите внимание на бейзлайн - прошлые покупки юзера)  

Сделайте MVP - минимально рабочий продукт - (пусть даже top-popular), а потом его улучшайте

Если вы делаете двухуровневую модель - следите за валидацией 

# Import libs

In [1]:
import pandas as pd
import numpy as np
import statistics
import matplotlib.pyplot as plt
%matplotlib inline

# Для работы с матрицами
from scipy.sparse import csr_matrix

# Матричная факторизация
from implicit import als

# Модель второго уровня
from lightgbm import LGBMClassifier

import os, sys
module_path = os.path.abspath(os.path.join(os.pardir))
if module_path not in sys.path:
    sys.path.append(module_path)

# Написанные нами функции
from metrics import precision_at_k, recall_at_k
from utils import prefilter_items
from recommenders import MainRecommender

## Read data

In [2]:
data = pd.read_csv('retail_train.csv')
item_features = pd.read_csv('product.csv')
user_features = pd.read_csv('hh_demographic.csv')

In [3]:
# data.head()

# Set global const

In [4]:
ITEM_COL = 'item_id'
USER_COL = 'user_id'
ACTUAL_COL = 'actual'

# N = Neighbors
N_PREDICT = 50 

# Process features dataset

In [5]:
# column processing
item_features.columns = [col.lower() for col in item_features.columns]
user_features.columns = [col.lower() for col in user_features.columns]

item_features.rename(columns={'product_id': ITEM_COL}, inplace=True)
user_features.rename(columns={'household_key': USER_COL }, inplace=True)

# Split dataset for train, eval, test

In [6]:
# Важна схема обучения и валидации!
# -- давние покупки -- | -- 6 недель -- | -- 3 недель -- 
# подобрать размер 2-ого датасета (6 недель) --> learning curve (зависимость метрики recall@k от размера датасета)


VAL_MATCHER_WEEKS = 6
VAL_RANKER_WEEKS = 3

In [7]:
# берем данные для тренировки matching модели
data_train_matcher = data[data['week_no'] < data['week_no'].max() - (VAL_MATCHER_WEEKS + VAL_RANKER_WEEKS)]

# берем данные для валидации matching модели
data_val_matcher = data[(data['week_no'] >= data['week_no'].max() - (VAL_MATCHER_WEEKS + VAL_RANKER_WEEKS)) &
                      (data['week_no'] < data['week_no'].max() - (VAL_RANKER_WEEKS))]

# берем данные для тренировки ranking модели
data_train_ranker = data_val_matcher.copy()  # Для наглядности. Далее мы добавим изменения, и они будут отличаться

# берем данные для теста ranking, matching модели
data_val_ranker = data[data['week_no'] >= data['week_no'].max() - VAL_RANKER_WEEKS]

In [8]:
# сделаем объединенный сет данных для первого уровня (матчинга)
df_join_train_matcher = pd.concat([data_train_matcher, data_val_matcher])

In [9]:
def print_stats_data(df_data, name_df):
    print(name_df)
    print(f"Shape: {df_data.shape} Users: {df_data[USER_COL].nunique()} Items: {df_data[ITEM_COL].nunique()}")

In [10]:
print_stats_data(data_train_matcher,'train_matcher')
print_stats_data(data_val_matcher,'val_matcher')
print_stats_data(data_train_ranker,'train_ranker')
print_stats_data(data_val_ranker,'val_ranker')

train_matcher
Shape: (2108779, 12) Users: 2498 Items: 83685
val_matcher
Shape: (169711, 12) Users: 2154 Items: 27649
train_ranker
Shape: (169711, 12) Users: 2154 Items: 27649
val_ranker
Shape: (118314, 12) Users: 2042 Items: 24329


выше видим разброс по пользователям и товарам и дальше мы перейдем к warm-start (только известные пользователи)

In [11]:
data_val_matcher.head(2)

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
2104867     2070  40618492260  594  1019940         1         1.00       311   
2107468     2021  40618753059  594   840361         1         0.99       443   

         retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
2104867        -0.29          40       86          0.0                0.0  
2107468         0.00         101       86          0.0                0.0

# Prefilter items

In [12]:
n_items_before = data_train_matcher['item_id'].nunique()

data_train_matcher = prefilter_items(data_train_matcher, item_features=item_features, take_n_popular=5000)

n_items_after = data_train_matcher['item_id'].nunique()
print('Decreased # items from {} to {}'.format(n_items_before, n_items_after))

C:\Users\Данила\Documents\Data science\RecSys\course_project\utils.py:20: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['price'] = data['sales_value'] / (np.maximum(data['quantity'], 1))


Decreased # items from 83685 to 5001


# Make cold-start to warm-start

In [13]:
# ищем общих пользователей
common_users = list(set(data_train_matcher.user_id.values)&(set(data_val_matcher.user_id.values))&set(data_val_ranker.user_id.values))

# оставляем общих пользователей
data_train_matcher = data_train_matcher[data_train_matcher.user_id.isin(common_users)]
data_val_matcher = data_val_matcher[data_val_matcher.user_id.isin(common_users)]
data_train_ranker = data_train_ranker[data_train_ranker.user_id.isin(common_users)]
data_val_ranker = data_val_ranker[data_val_ranker.user_id.isin(common_users)]

print_stats_data(data_train_matcher,'train_matcher')
print_stats_data(data_val_matcher,'val_matcher')
print_stats_data(data_train_ranker,'train_ranker')
print_stats_data(data_val_ranker,'val_ranker')

train_matcher
Shape: (784420, 13) Users: 1915 Items: 4999
val_matcher
Shape: (163261, 12) Users: 1915 Items: 27118
train_ranker
Shape: (163261, 12) Users: 1915 Items: 27118
val_ranker
Shape: (115989, 12) Users: 1915 Items: 24042


# Init/train recommender

In [14]:
recommender = MainRecommender(data_train_matcher)

# Модель первого уровня (matching)

In [15]:
result_eval_matcher = data_val_matcher.groupby(USER_COL)[ITEM_COL].unique().reset_index()
result_eval_matcher.columns=[USER_COL, ACTUAL_COL]
result_eval_matcher.head(2)

user_id                                             actual
0        1  [853529, 865456, 867607, 872137, 874905, 87524...
1        6  [1024306, 1102949, 6548453, 835394, 940804, 96...

In [16]:
result_eval_matcher['own_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))
result_eval_matcher['sim_item_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_similar_items_recommendation(x, N=N_PREDICT))
result_eval_matcher['als_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_als_recommendations(x, N=N_PREDICT))
# result_eval_matcher['sim_user_rec'] = result_eval_matcher[USER_COL].apply(lambda x: recommender.get_similar_users_recommendation(x, N=N_PREDICT))
result_eval_matcher.head()

user_id                                             actual  \
0        1  [853529, 865456, 867607, 872137, 874905, 87524...   
1        6  [1024306, 1102949, 6548453, 835394, 940804, 96...   
2        7  [836281, 843306, 845294, 914190, 920456, 93886...   
3        8  [868075, 886787, 945611, 1005186, 1008787, 101...   
4        9  [883616, 1029743, 1039126, 1051323, 1082772, 1...   

                                             own_rec  \
0  [856942, 9297615, 5577022, 877391, 9655212, 10...   
1  [13003092, 995598, 923600, 972416, 1084036, 11...   
2  [998519, 894360, 7147142, 9338009, 896666, 939...   
3  [12808385, 939860, 981660, 7410201, 5577022, 6...   
4  [872146, 918046, 9655676, 985622, 1056005, 109...   

                                        sim_item_rec  \
0  [842762, 12301405, 826120, 5577022, 948384, 98...   
1  [948650, 5569845, 8357613, 941361, 1074754, 11...   
2  [5585510, 7147145, 1044078, 880427, 948468, 83...   
3  [5569845, 5569374, 1044078, 1128016, 995358, 8...   
4  [1008032, 1074754, 901062, 982537, 898363, 713...   

                                             als_rec  
0  [6533936, 856942, 883616, 888543, 824758, 1047...  
1  [1026118, 854852, 1084036, 965267, 970160, 102...  
2  [912451, 1039627, 974250, 7142937, 851287, 912...  
3  [860248, 916122, 12352054, 831888, 998236, 830...  
4  [8205418, 5585510, 1029743, 6463710, 8090541, ...

In [17]:
def calc_recall(df_data, top_k):
    for col_name in df_data.columns[2:]:
        yield col_name, df_data.apply(lambda row: recall_at_k(row[col_name], row[ACTUAL_COL], k=top_k), axis=1).mean()

In [18]:
def calc_precision(df_data, top_k):
    for col_name in df_data.columns[2:]:
        yield col_name, df_data.apply(lambda row: precision_at_k(row[col_name], row[ACTUAL_COL], k=top_k), axis=1).mean()

### Recall@50 of matching

In [19]:
TOPK_RECALL = 50

In [20]:
sorted(calc_recall(result_eval_matcher, TOPK_RECALL), key=lambda x: x[1],reverse=True)

[('own_rec', 0.061684201353290766),
 ('als_rec', 0.04683256067563649),
 ('sim_item_rec', 0.03216272431160577)]

### Precision@5 of matching

In [21]:
TOPK_PRECISION = 5

In [22]:
sorted(calc_precision(result_eval_matcher, TOPK_PRECISION), key=lambda x: x[1],reverse=True)

[('own_rec', 0.18872062663185182),
 ('als_rec', 0.12167101827676158),
 ('sim_item_rec', 0.0679895561357707)]

Наилучшие метрики у модели own_recommendations, поэтому для будем использовать ее

# Ranking part

### Обучаем модель 2-ого уровня на выбранных кандидатах

- Обучаем на data_train_ranking
- Обучаем *только* на выбранных кандидатах
- Я *для примера* сгенерирую топ-50 кадидиатов через get_own_recommendations
- (!) Если юзер купил < 50 товаров, то get_own_recommendations дополнит рекоммендации топ-популярными

In [23]:
# -- давние покупки -- | -- 6 недель -- | -- 3 недель -- 

## Подготовка данных для трейна

In [24]:
# взяли пользователей из трейна для ранжирования
df_match_candidates = pd.DataFrame(data_train_ranker[USER_COL].unique())
df_match_candidates.columns = [USER_COL]

In [25]:
# собираем кандитатов с первого этапа (matcher)
df_match_candidates['candidates'] = df_match_candidates[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))

In [26]:
df_match_candidates.head(2)

user_id                                         candidates
0     2070  [1105426, 1097350, 879194, 948640, 928263, 944...
1     2021  [950935, 1119454, 835578, 863762, 1097398, 101...

In [27]:
# разворачиваем товары
df_items = df_match_candidates.apply(lambda x: pd.Series(x['candidates']), axis=1).stack().reset_index(level=1, drop=True)
df_items.name = 'item_id'

In [28]:
df_match_candidates = df_match_candidates.drop('candidates', axis=1).join(df_items)

In [29]:
df_match_candidates.head(4)

user_id  item_id
0     2070  1105426
0     2070  1097350
0     2070   879194
0     2070   948640

### Check warm start

In [30]:
print_stats_data(df_match_candidates, 'match_candidates')

match_candidates
Shape: (95750, 2) Users: 1915 Items: 4437


Количество уникальных пользователей не изменилось

### Создаем трейн сет для ранжирования с учетом кандидатов с 1-го этапа

In [31]:
df_ranker_train = data_train_ranker[[USER_COL, ITEM_COL]].copy()
df_ranker_train['target'] = 1  # тут только покупки 

df_ranker_train = df_match_candidates.merge(df_ranker_train, on=[USER_COL, ITEM_COL], how='left')

df_ranker_train['target'].fillna(0, inplace= True)

In [32]:
df_ranker_train.target.value_counts()

0.0    88346
1.0    11053
Name: target, dtype: int64

In [33]:
df_ranker_train.head(9)

user_id  item_id  target
0     2070  1105426     0.0
1     2070  1097350     0.0
2     2070   879194     0.0
3     2070   948640     0.0
4     2070   928263     0.0
5     2070   944588     0.0
6     2070  1032703     0.0
7     2070  1138596     0.0
8     2070  1092937     1.0

(!) На каждого юзера 50 item_id-кандидатов

In [34]:
df_ranker_train['target'].mean()

0.11119830179378062

## Подготавливаем фичи для обучения модели

### Описательные фичи

In [35]:
item_features.head(2)

item_id  manufacturer    department     brand            commodity_desc  \
0    25671             2       GROCERY  National                  FRZN ICE   
1    26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION

In [36]:
user_features.head(2)

age_desc marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   

  household_size_desc kid_category_desc  user_id  
0                   2      None/Unknown        1  
1                   2      None/Unknown        7

In [37]:
df_ranker_train = df_ranker_train.merge(item_features, on='item_id', how='left')
df_ranker_train = df_ranker_train.merge(user_features, on='user_id', how='left')

df_ranker_train.head(2)

user_id  item_id  target  manufacturer department     brand commodity_desc  \
0     2070  1105426     0.0            69       DELI   Private     SANDWICHES   
1     2070  1097350     0.0          2468    GROCERY  National  DOMESTIC WINE   

    sub_commodity_desc curr_size_of_product age_desc marital_status_code  \
0  SANDWICHES - (COLD)                         45-54                   U   
1     VALUE GLASS WINE                4 LTR    45-54                   U   

  income_desc homeowner_desc hh_comp_desc household_size_desc  \
0      50-74K        Unknown      Unknown                   1   
1      50-74K        Unknown      Unknown                   1   

  kid_category_desc  
0      None/Unknown  
1      None/Unknown

**Фичи user_id:**

    - Средний чек
    - Средняя сумма покупки 1 товара в каждой категории
    - Кол-во покупок в каждой категории
    - Частотность покупок раз/месяц
    - Долю покупок в выходные
    - Долю покупок утром/днем/вечером

**Фичи item_id**:

    - Кол-во покупок в неделю
    - Среднее ол-во покупок 1 товара в категории в неделю
    - (Кол-во покупок в неделю) / (Среднее кол-во покупок 1 товара в категории в неделю)
    - Цена (Можно посчитать из retil_train.csv)
    - Цена / Средняя цена товара в категории
    
**Фичи пары user_id - item_id**

    - (Средняя сумма покупки 1 товара в каждой категории (берем категорию item_id)) - (Цена item_id)
    - (Кол-во покупок юзером конкретной категории в неделю) - (Среднее кол-во покупок всеми юзерами конкретной категории в неделю)
    - (Кол-во покупок юзером конкретной категории в неделю) / (Среднее кол-во покупок всеми юзерами конкретной категории в неделю)

### Поведенческие фичи

##### Чтобы считать поведенческие фичи, нужно учесть все данные что были до data_val_ranker

In [38]:
df_join_train_matcher.head()

user_id    basket_id  day  item_id  quantity  sales_value  store_id  \
0     2375  26984851472    1  1004906         1         1.39       364   
1     2375  26984851472    1  1033142         1         0.82       364   
2     2375  26984851472    1  1036325         1         0.99       364   
3     2375  26984851472    1  1082185         1         1.21       364   
4     2375  26984851472    1  8160430         1         1.50       364   

   retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0        -0.60        1631        1          0.0                0.0  
1         0.00        1631        1          0.0                0.0  
2        -0.30        1631        1          0.0                0.0  
3         0.00        1631        1          0.0                0.0  
4        -0.39        1631        1          0.0                0.0

#Фичи для user_id
def new_user_features(data, user_features, item_features):
    
    new_user_features = user_features.merge(data, on='user_id', how='left')
    new_item_features = data.merge(item_features, on='item_id', how='left')
    
    #Средний чек
    average_check = new_user_features.groupby(['user_id'])['sales_value'].mean().reset_index()
    average_check.rename(columns={'sales_value': 'average_sales_value'}, inplace=True)
    user_features = user_features.merge(average_check)

     
    #Количество уникальных категорий покупателя
    num_unique_department = new_item_features.groupby(['user_id'])['department'].nunique().reset_index()
    num_unique_department.rename(columns={'department': 'num_unique_department'}, inplace=True)
    user_features = user_features.merge(num_unique_department, on='user_id', how='left')
    
   
    #Количество покупок в каждой категории
    num_sales_in_category = new_item_features[['user_id', 'quantity', 'department']].groupby(['user_id', 'department']).sum().reset_index()
    num_sales_in_category.rename(columns={'quantity': 'num_quantity'}, inplace=True) 
    num_sales_in_category.drop(labels=[0], axis=0,inplace=True)

    users_num_cat_sales_dict = {}
    for user_id in num_sales_in_category['user_id'].unique():
        users_num_cat_sales_dict[user_id] = dict(list(zip(num_sales_in_category[num_sales_in_category['user_id']==user_id]['department'].values, \
                                                      num_sales_in_category[num_sales_in_category['user_id']==user_id]['num_quantity'].values)))

    for cat in item_features['department'].unique():
        new_item_features[f'num_sales_in_{cat}'] = 0

    for user_id in new_item_features['user_id'].unique():
        for cat in users_num_cat_sales_dict[user_id].keys():
            new_item_features.loc[(new_item_features['user_id']==user_id) & (new_item_features['department']==cat), f'num_sales_in_{cat}']=\
            users_num_cat_sales_dict[user_id][cat]

    feat_to_merge = ['user_id'] + new_item_features.columns.tolist()[18:]
    user_num_sales_in_cat = new_item_features[feat_to_merge]
    user_num_sales_in_cat = user_num_sales_in_cat.groupby('user_id').max().reset_index()
    user_features = user_features.merge(user_num_sales_in_cat, on='user_id', how='left')
    
    user_features = user_features.drop('num_sales_in_ ', axis=1)
    
    #Среднее время покупки пользователем
    new_user_features['hour'] = new_user_features['trans_time'] // 100
    hour = new_user_features.groupby('user_id')['hour'].median().reset_index()
    hour.columns = ['user_id', 'median_sales_hour_for_user']
    user_features = user_features.merge(hour, on='user_id', how='left')
    
    #Популярный день недели покупки юзера
    new_user_features['weekday'] = new_user_features['day'] % 7
    week_day = new_user_features.groupby('user_id')['weekday'].apply(statistics.mode).reset_index()
    week_day.columns = ['user_id', 'mode_sales_day_for_user']
    user_features = user_features.merge(week_day, on='user_id', how='left')
    
    # средний чек корзины покупателя
    average_check_bascket = new_user_features.groupby(['user_id', 'basket_id'])['sales_value'].sum().reset_index()
    average_check_bascket = average_check_bascket.groupby('user_id')['sales_value'].mean().reset_index()
    average_check_bascket.columns = ['user_id', 'mean_basket_check']
    user_features = user_features.merge(average_check_bascket, on='user_id', how='left')
    
    # кол-во уникальных товаров, купленных клиентом
    item_unique = new_user_features.groupby(['user_id'])['item_id'].nunique().reset_index()
    item_unique.columns = ['user_id', 'n_items']
    user_features = user_features.merge(item_unique, on='user_id', how='left')
    
    user_features = user_features.replace(np.nan, 0)
    
    return user_features

user_features_new = new_user_features(data_train_ranker, user_features, item_features)
user_features_new.head(3)

In [39]:
def new_item_features(data, user_features, item_features):
    
    new_item_features = item_features.merge(data, on='item_id', how='left')
    new_user_features = user_features.merge(data, on='user_id', how='left')
    
    # Цена
    price = new_item_features.groupby('item_id')['sales_value'].sum() \
                            / new_item_features.groupby('item_id')['quantity'].sum()
    price = price.groupby('item_id').mean().reset_index()
    price.columns = ['item_id', 'price']
    price['price'].fillna(0, inplace= True)
    item_features = item_features.merge(price)

    #Среднее число покупок товара в неделю
    num_purchase_week = data.groupby(['item_id']).agg({'week_no': 'nunique', 'quantity': 'sum'}).reset_index()
    num_purchase_week['average_num_purchases_week'] = num_purchase_week['quantity'] / num_purchase_week['week_no']
    average_num_purchases_week = num_purchase_week[['item_id', 'average_num_purchases_week']]
    item_features = item_features.merge(average_num_purchases_week, on='item_id', how='left')
    item_features['average_num_purchases_week'].fillna(0, inplace= True)
    
    #Средняя цена товара в категории
    average_price_in_cat = item_features.groupby(['department'])['price'].mean().reset_index()
    average_price_in_cat.rename(columns={'price': 'average_price_in_dep'}, inplace=True)
    average_price_in_cat.drop(labels=[0], axis=0,inplace=True)
    item_features = item_features.merge(average_price_in_cat)
    
    #Цена товара/средняя цена товара в категории
    item_features['price/average_price_in_dep'] = item_features['price'] / item_features['average_price_in_dep']
    
    #Средняя цена товара в категории - Цена товара
    item_features['average_price_in_dep - item_price'] = item_features['average_price_in_dep'] - item_features['price']
    
    #Среднее время покупки товара
    new_item_features['hour'] = new_item_features['trans_time'] // 100
    hour = new_item_features.groupby('item_id')['hour'].median().reset_index()
    hour.columns = ['item_id', 'median_sales_hour_for_item']
    item_features = item_features.merge(hour, on='item_id', how='left')
    
    #Популярный день недели покупки товара
    new_item_features['weekday'] = new_item_features['day'] % 7
    week_day = new_item_features.groupby('item_id')['weekday'].apply(statistics.mode).reset_index()
    week_day.columns = ['item_id', 'mode_sales_day_for_item']
    item_features = item_features.merge(week_day, on='item_id', how='left')
    
    #Количество магазинов, в которых продается товар
    shops = new_item_features.groupby('item_id')['store_id'].nunique().reset_index()
    shops.columns = ['item_id', 'n_stores']
    item_features = item_features.merge(shops, on='item_id', how='left')
    
    item_features = item_features.replace(np.nan, 0)
    
    return item_features

In [40]:
item_features_new = new_item_features(data_train_ranker, user_features, item_features)
item_features_new.head(3)

item_id  manufacturer department     brand        commodity_desc  \
0    25671             2    GROCERY  National              FRZN ICE   
1    26190            69    GROCERY   Private  FRUIT - SHELF STABLE   
2    26355            69    GROCERY   Private         COOKIES/CONES   

    sub_commodity_desc curr_size_of_product  price  \
0  ICE - CRUSHED/CUBED                22 LB    0.0   
1          APPLE SAUCE                50 OZ    0.0   
2    SPECIALTY COOKIES                14 OZ    0.0   

   average_num_purchases_week  average_price_in_dep  \
0                         0.0              1.043754   
1                         0.0              1.043754   
2                         0.0              1.043754   

   price/average_price_in_dep  average_price_in_dep - item_price  \
0                         0.0                           1.043754   
1                         0.0                           1.043754   
2                         0.0                           1.043754   

   median_sales_hour_for_item  mode_sales_day_for_item  n_stores  
0                         0.0                      0.0         0  
1                         0.0                      0.0         0  
2                         0.0                      0.0         0

In [41]:
df_ranker_train = df_ranker_train.merge(item_features_new, on='item_id', how='left')
# df_ranker_train = df_ranker_train.merge(user_features_new, on='user_id', how='left')

## !!! Пока выполните нотбук без этих строк, потом вернитесь и запустите их, обучите ранкер и посмотрите на метрики с ранжированием

In [42]:
# df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg('sales_value').sum().rename('total_item_sales_value'), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg('quantity').sum().rename('total_quantity_value'), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg(USER_COL).count().rename('item_freq'), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=USER_COL).agg(USER_COL).count().rename('user_freq'), how='left',on=USER_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=USER_COL).agg('sales_value').sum().rename('total_user_sales_value'), how='left',on=USER_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg('quantity').sum().rename('item_quantity_per_week')/df_join_train_matcher.week_no.nunique(), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=USER_COL).agg('quantity').sum().rename('user_quantity_per_week')/df_join_train_matcher.week_no.nunique(), how='left',on=USER_COL)


df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg('quantity').sum().rename('item_quantity_per_basket')/df_join_train_matcher.basket_id.nunique(), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=USER_COL).agg('quantity').sum().rename('user_quantity_per_baskter')/df_join_train_matcher.basket_id.nunique(), how='left',on=USER_COL)


df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=ITEM_COL).agg(USER_COL).count().rename('item_freq_per_basket')/df_join_train_matcher.basket_id.nunique(), how='left',on=ITEM_COL)

df_ranker_train = df_ranker_train.merge(df_join_train_matcher.groupby(by=USER_COL).agg(USER_COL).count().rename('user_freq_per_basket')/df_join_train_matcher.basket_id.nunique(), how='left',on=USER_COL)

In [43]:
df_ranker_train.head()

user_id  item_id  target  manufacturer_x department_x   brand_x  \
0     2070  1105426     0.0              69         DELI   Private   
1     2070  1097350     0.0            2468      GROCERY  National   
2     2070   879194     0.0              69      DRUG GM   Private   
3     2070   948640     0.0            1213      DRUG GM  National   
4     2070   928263     0.0              69      DRUG GM   Private   

        commodity_desc_x sub_commodity_desc_x curr_size_of_product_x age_desc  \
0             SANDWICHES  SANDWICHES - (COLD)                           45-54   
1          DOMESTIC WINE     VALUE GLASS WINE                  4 LTR    45-54   
2  DIAPERS & DISPOSABLES         BABY DIAPERS                  14 CT    45-54   
3  ORAL HYGIENE PRODUCTS    WHITENING SYSTEMS                   3 OZ    45-54   
4  DIAPERS & DISPOSABLES         BABY DIAPERS                  13 CT    45-54   

   ... total_quantity_value item_freq user_freq total_user_sales_value  \
0  ...                  113        99      1996                5754.86   
1  ...                   54        51      1996                5754.86   
2  ...                   54        46      1996                5754.86   
3  ...                   49        44      1996                5754.86   
4  ...                   59        53      1996                5754.86   

  item_quantity_per_week user_quantity_per_week  item_quantity_per_basket  \
0               1.241758             1218.32967                  0.000461   
1               0.593407             1218.32967                  0.000220   
2               0.593407             1218.32967                  0.000220   
3               0.538462             1218.32967                  0.000200   
4               0.648352             1218.32967                  0.000241   

  user_quantity_per_baskter item_freq_per_basket user_freq_per_basket  
0                  0.452137             0.000404              0.00814  
1                  0.452137             0.000208              0.00814  
2                  0.452137             0.000188              0.00814  
3                  0.452137             0.000179              0.00814  
4                  0.452137             0.000216              0.00814  

[5 rows x 40 columns]

In [44]:
X_train = df_ranker_train.drop('target', axis=1)
y_train = df_ranker_train[['target']]

In [45]:
cat_feats = X_train.columns[2:].tolist()

X_train[cat_feats] = X_train[cat_feats].astype('category')

cat_feats

['manufacturer_x',
 'department_x',
 'brand_x',
 'commodity_desc_x',
 'sub_commodity_desc_x',
 'curr_size_of_product_x',
 'age_desc',
 'marital_status_code',
 'income_desc',
 'homeowner_desc',
 'hh_comp_desc',
 'household_size_desc',
 'kid_category_desc',
 'manufacturer_y',
 'department_y',
 'brand_y',
 'commodity_desc_y',
 'sub_commodity_desc_y',
 'curr_size_of_product_y',
 'price',
 'average_num_purchases_week',
 'average_price_in_dep',
 'price/average_price_in_dep',
 'average_price_in_dep - item_price',
 'median_sales_hour_for_item',
 'mode_sales_day_for_item',
 'n_stores',
 'total_quantity_value',
 'item_freq',
 'user_freq',
 'total_user_sales_value',
 'item_quantity_per_week',
 'user_quantity_per_week',
 'item_quantity_per_basket',
 'user_quantity_per_baskter',
 'item_freq_per_basket',
 'user_freq_per_basket']

## Обучение модели ранжирования

In [46]:
lgb = LGBMClassifier(objective='binary',
                     max_depth=10,
                     n_estimators=500,
                     learning_rate=0.05,
                     categorical_column=cat_feats)

lgb.fit(X_train, y_train)

train_preds = lgb.predict_proba(X_train)

C:\Users\Данила\AppData\Roaming\Python\Python38\site-packages\sklearn\utils\validation.py:63: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(*args, **kwargs)
C:\ProgramData\Anaconda3\lib\site-packages\lightgbm\basic.py:1245: UserWarning: categorical_column in param dict is overridden.
  _log_warning('{} in param dict is overridden.'.format(cat_alias))


In [47]:
df_ranker_predict = df_ranker_train.copy()

In [48]:
df_ranker_predict['proba_item_purchase'] = train_preds[:,1]

In [49]:
df_ranker_predict

user_id  item_id  target  manufacturer_x department_x   brand_x  \
0         2070  1105426     0.0              69         DELI   Private   
1         2070  1097350     0.0            2468      GROCERY  National   
2         2070   879194     0.0              69      DRUG GM   Private   
3         2070   948640     0.0            1213      DRUG GM  National   
4         2070   928263     0.0              69      DRUG GM   Private   
...        ...      ...     ...             ...          ...       ...   
99394     1745   948832     0.0            1719   MEAT-PCKGD  National   
99395     1745   903454     0.0            1216   MEAT-PCKGD  National   
99396     1745  9419888     0.0             759      GROCERY  National   
99397     1745   880469     0.0             544      GROCERY  National   
99398     1745  1024051     0.0            1508      GROCERY  National   

             commodity_desc_x          sub_commodity_desc_x  \
0                  SANDWICHES           SANDWICHES - (COLD)   
1               DOMESTIC WINE              VALUE GLASS WINE   
2       DIAPERS & DISPOSABLES                  BABY DIAPERS   
3       ORAL HYGIENE PRODUCTS             WHITENING SYSTEMS   
4       DIAPERS & DISPOSABLES                  BABY DIAPERS   
...                       ...                           ...   
99394                HOT DOGS                BETTER FOR YOU   
99395             FROZEN MEAT          OTHER - FULLY COOKED   
99396                  YOGURT            YOGURT MULTI-PACKS   
99397              BAG SNACKS                    CORN CHIPS   
99398  FRZN MEAT/MEAT DINNERS  FRZN MULTI SERVE ENTREES ALL   

      curr_size_of_product_x age_desc  ... item_freq user_freq  \
0                               45-54  ...        99      1996   
1                      4 LTR    45-54  ...        51      1996   
2                      14 CT    45-54  ...        46      1996   
3                       3 OZ    45-54  ...        44      1996   
4                      13 CT    45-54  ...        53      1996   
...                      ...      ...  ...       ...       ...   
99394                   1 LB    45-54  ...       252       897   
99395                  32 OZ    45-54  ...        48       897   
99396                  48 OZ    45-54  ...        87       897   
99397                  10 OZ    45-54  ...       172       897   
99398                  28 OZ    45-54  ...       237       897   

      total_user_sales_value item_quantity_per_week user_quantity_per_week  \
0                    5754.86               1.241758            1218.329670   
1                    5754.86               0.593407            1218.329670   
2                    5754.86               0.593407            1218.329670   
3                    5754.86               0.538462            1218.329670   
4                    5754.86               0.648352            1218.329670   
...                      ...                    ...                    ...   
99394                2397.25               3.967033              13.208791   
99395                2397.25               0.549451              13.208791   
99396                2397.25               1.000000              13.208791   
99397                2397.25               1.945055              13.208791   
99398                2397.25               3.769231              13.208791   

      item_quantity_per_basket  user_quantity_per_baskter  \
0                     0.000461                   0.452137   
1                     0.000220                   0.452137   
2                     0.000220                   0.452137   
3                     0.000200                   0.452137   
4                     0.000241                   0.452137   
...                        ...                        ...   
99394                 0.001472                   0.004902   
99395                 0.000204                   0.004902   
99396                 0.000371                   0.004902   
99397                

## Подведем итоги

    Мы обучили модель ранжирования на покупках из сета data_train_ranker и на кандитатах от own_recommendations, что является тренировочным сетом, и теперь наша задача предсказать и оценить именно на тестовом сете.

# Evaluation on test dataset

In [50]:
result_eval_ranker = data_val_ranker.groupby(USER_COL)[ITEM_COL].unique().reset_index()
result_eval_ranker.columns=[USER_COL, ACTUAL_COL]
result_eval_ranker.head(2)

user_id                                             actual
0        1  [821867, 834484, 856942, 865456, 889248, 90795...
1        6  [920308, 926804, 946489, 1006718, 1017061, 107...

## Eval matching on test dataset

In [51]:
%%time
result_eval_ranker['own_rec'] = result_eval_ranker[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))

Wall time: 5.57 s


In [52]:
# померяем precision только модели матчинга, чтобы понимать влияение ранжирования на метрики

sorted(calc_precision(result_eval_ranker, TOPK_PRECISION), key=lambda x: x[1], reverse=True)

[('own_rec', 0.1462140992167092)]

## Eval re-ranked matched result on test dataset
    Вспомним df_match_candidates сет, который был получен own_recommendations на юзерах, набор пользователей мы фиксировали и он одинаков, значит и прогноз одинаков, поэтому мы можем использовать этот датафрейм для переранжирования.    

In [53]:
def rerank(data, user_id):
    return data[data[USER_COL]==user_id].sort_values('proba_item_purchase', ascending=False).head(5).item_id.tolist()

In [54]:
result_eval_ranker['reranked_own_rec'] = result_eval_ranker[USER_COL].apply(lambda user_id: rerank(df_ranker_predict, user_id))

In [55]:
print(*sorted(calc_precision(result_eval_ranker, TOPK_PRECISION), key=lambda x: x[1], reverse=True), sep='\n')

('reranked_own_rec', 0.1523759791122698)
('own_rec', 0.1462140992167092)


('own_rec', 0.1462140992167092)

без фичей пользователя и товара: ('reranked_own_rec', 0.15007832898172166)

без фичей товара: ('reranked_own_rec', 0.15112271540469807)

без фичей пользователя: ('reranked_own_rec', 0.1523759791122698)

все фичи: ('reranked_own_rec', 0.1507049608355075)

смотрим на метрики выше и сравниваем что с ранжированием и без, добавляем фичи и то же смотрим

# Оценка на тесте для выполнения курсового проекта

In [56]:
df_test = pd.read_csv('retail_test1.csv')
df_test = df_test[df_test[USER_COL].isin(common_users)]
df_transactions = pd.read_csv('retail_train.csv')

In [57]:
print_stats_data(df_test, 'df_test')

df_test
Shape: (83656, 12) Users: 1663 Items: 19981


In [58]:
result_test = df_test.groupby(USER_COL)[ITEM_COL].unique().reset_index()
result_test.columns=[USER_COL, ACTUAL_COL]
result_test.head(2)

user_id                                             actual
0        1  [880007, 883616, 931136, 938004, 940947, 94726...
1        6  [956902, 960791, 1037863, 1119051, 1137688, 84...

In [59]:
result_test['own_rec'] = result_test[USER_COL].apply(lambda x: recommender.get_own_recommendations(x, N=N_PREDICT))

sorted(calc_precision(result_test, TOPK_PRECISION), key=lambda x: x[1],reverse=True)

[('own_rec', 0.12928442573661988)]

In [61]:
result_test['reranked_own_rec'] = result_test[USER_COL].apply(lambda user_id: rerank(df_ranker_predict, user_id))

In [62]:
print(*sorted(calc_precision(result_test, TOPK_PRECISION), key=lambda x: x[1], reverse=True), sep='\n')

('reranked_own_rec', 0.13156945279615076)
('own_rec', 0.12928442573661988)


**Вывод:** на тестовых данных метрика precision@5 равна 0.13156945279615076